include/internal/refcount.h

```diff
@@ -193,14 +189,91 @@ static __inline int CRYPTO_GET_REF(volatile int *val, int *ret,
  */
 # ifndef HAVE_ATOMICS
 
-typedef int CRYPTO_REF_COUNT;
+typedef struct {
+    int val;
+#  ifdef OPENSSL_THREADS
+    CRYPTO_RWLOCK *lock;
+#  endif
+} CRYPTO_REF_COUNT;
+
+#  ifdef OPENSSL_THREADS
+
+static ossl_unused ossl_inline int CRYPTO_UP_REF(CRYPTO_REF_COUNT *refcnt, 
+                                                 int *ret)
+{
+    return CRYPTO_atomic_add(&refcnt->val, 1, ret, refcnt->lock);
+}
+
+static ossl_unused ossl_inline int CRYPTO_DOWN_REF(CRYPTO_REF_COUNT *refcnt,
+                                                   int *ret)
+{
+    return CRYPTO_atomic_add(&refcnt->val, -1, ret, refcnt->lock);
+}
+
+static ossl_unused ossl_inline int CRYPTO_GET_REF(CRYPTO_REF_COUNT *refcnt,
+                                                   int *ret)
+{
+    return CRYPTO_atomic_load_int(&refcnt->val, ret, refcnt->lock);
+}
+
+#   define CRYPTO_NEW_FREE_DEFINED  1
+static ossl_unused ossl_inline int CRYPTO_NEW_REF(CRYPTO_REF_COUNT *refcnt, int n)
+{
+    refcnt->val = n;
+    refcnt->lock = CRYPTO_THREAD_lock_new();
+    if (refcnt->lock == NULL) {
+        ERR_raise(ERR_LIB_CRYPTO, ERR_R_CRYPTO_LIB);
+        return 0;
+    }
+    return 1;
+}
+
+static ossl_unused ossl_inline void CRYPTO_FREE_REF(CRYPTO_REF_COUNT *refcnt)                                  \
+{
+    CRYPTO_THREAD_lock_free(refcnt->lock);
```

Comment:
This function should support refcnt == NULL. The code changes actually make it possible to call this on error with it still being NULL.

Reply:
I don't think this is needed.  The argument is a pointer to a field inside a structure and thus cannot ever be null.
Still, the check should be harmless and ought to optimise out.

Reply:
You're right, it's something that is always already allocated, refcnt should never be NULL.